src/smtpc/smtpc.1
.TH SMTPC 1 "17 May 2015" 0.1a.1 "smtpc 0.1a.1"

.SH NAME
smtpc \- SMTP / LMTP client

.SH SYNOPSIS
smtpc --esmtp \fIhost\fR:\fIport\fR -f \fIenvelope@sen.der\fR
[ \fIoptions\fR... ] [ -- ] \fIrecipient@add.ress\fR ...

smtpc --lmtp \fIpath\fR -f \fIenvelope@sen.der\fR
[ \fIoptions\fR... ] [ -- ] \fIrecipient@add.ress\fR ...

.SH DESCRIPTION
\fBsmtpc\fP is a mail client.
It reads a message from standart input and submits it to mail server through
socket.

.SS Options
Any options not listed here are simply ignored.

.TP
\fB--dump\fP
Show dialog in the session.

.TP
\fB--esmtp\fP \fIhost\fR[:\fIport\fR]
Use TCP socket and ESMTP protocol to submit message.
Either this option or \fB--lmtp\fP option is required.

If \fIhost\fR is the IPv6 address, it must be enclosed in [...]
to avoid confusion with colon separating \fIhost\fR and \fIport\fR,
e.g. "[::1]".

.TP
\fB-f\fP \fIenvelope@sen.der\fR, \fB-f\fP\fIenvelope@sen.der\fR
Specifys envelope sender.
This option is required.

To specify "null envelope sender", use a separate empty argument or "<>".

.TP
\fB--iam\fP \fIhost.name\fR
Specifys host name or IP address literal used in EHLO request.
Default is "localhost".

.TP
\fB--lmtp\fP \fIpath\fR
Use Unix domain socket and LMTP protocol to submit message.
Either this option or \fB--esmtp\fP option is required.

.TP
\fB-N\fP \fIdsn\fR, \fB-N\fP\fIdsn\fR
Controls delivery status notification.
\fIdsn\fR may be single word "NEVER" or one or more of words "SUCCESS",
"FAILURE" and "DELAY" separated by comma.

If this option is not given, delivery status notification will be controlled
by server.

.TP
\fB--smtputf8\fP
Enables support for SMTPUTF8 extension.
\fBsmtpc\fR detects valid UTF-8 sequence in envelope and message header,
then requests this extension as neccessity.

.TP
\fB-V\fP \fIenvid\fR, \fB-N\fP\fIenvid\fR
Specifys envelope ID.

.TP
\fB--verbose\fP
Output the last response from the server to standard output.

.TP
\fB--\fP
Ends options.
Remainder of command line arguments are concidered to be recipient addresses.

.TP
\fIrecipent@add.ress\fR ...
Recipients to whom the message would be delivered.

.SS Exit status

.TP
\fB0\fR
Message was successfully submitted.

.TP
\fB253\fR
Message was rejected by server.

.TP
\fB254\fR
The server returns malformed or illegal response.

.TP
\fB255\fR
Network error occurred.

.SS "SMTP extensions"

\fBsmtpc\fR supports following extensions.

.TP
\fB8-bit MIME Transport\fR (RFC 6152)
\fBsmtpc\fR requests this extension if message contains octets with high bit.

.TP
\fBDelivery Status Notification\fR (RFC 3461)
See \fB-N\fR and \fB-V\fR options.

.TP
\fBMessage Size Declaration\fR (RFC 1870)
Estimated size of the message is informed to the server.

.TP
\fBInternationalized Email\fR (RFC 6531)
Experimentally supported.
See \fB--smtputf8\fR option.

.SH LIMITATIONS

\fBsmtpc\fR provides the feature of SMTP / LMTP client submitting messages
to particular server.
It will never provide extensive feature such as message queuing, retry after
temporary failure, mail routing using MX DNS record and so on.
Once the server rejects delivery, \fBsmtpc\fR exits and message is discarded.

.SH KNOWN BUGS

.TP 2
*
If NUL octets (\\0) are included in messages, they are transmitted to the
server.

.SH "SEE ALSO"
sendmail(1)

.SH HISTORY
\fBsmtpc\fP was initially written for Sympa project by
IKEDA Soji <ikeda@conversion.co.jp>.



NEWS
*****    Changes which are not compatible with previous versions are
*****    listed using "*****"in the margin. Please check this at least.

(See Changelog file for the full changes log)

6.2b.9		March 13, 2015

  Features:

	[12050] src/lib/Sympa/DatabaseDriver/Oracle.pm:  Now Oracle driver can
	check database structure.
	However, you still must update structure manually.

  Bug fixes:

	[12059] default/web_tt2/d_control.tt2, src/cgi/wwsympa.fcgi.in:  Sympa
	would issue an error when trying to edit access control to a file in
	the shared repository. This was due to a mismatch in parameters between
	list owenrs and document owners.

	[12058] src/lib/Conf.pm: [Reported by P. Grzesina, univ. of Regina]
	Sympa issued errors when trying to setupe a CAS server for SSO login.
	This was due to Sympa setting a hash value to scalar and then trying to
	use this value as a hash ref. Fixed by deleting affectation to scalar
	value - which is immediately changed to a hash anyway.

	[12055] src/cgi/wwsympa.fcgi.in, src/sbin/sympa.pl.in:	wwsympa and
	sympa/health_check did not check change of "cookie" parameter.

	[12053] default/web_tt2/list_menu.tt2:	Additional list menu links had
	vanished when setting up the new skin.

	[12052] src/lib/Sympa/List.pm, src/sbin/archived.pl.in,
	src/sbin/bounced.pl.in, src/sbin/bulk.pl.in,
	src/sbin/sympa_automatic.pl.in, src/sbin/sympa_msg.pl.in,
	src/sbin/task_manager.pl.in:  (probably fixes bug[#6204])
	task_manager.pl sometimes ignores signal to shutdown it.  If there are
	large number of lists, calling get_lists() can take long time and the
	process cannot deal with TERM signal.
	Now get_lists() checks HUP, INT and TERM signals and returns
	immediately when signal is raised.  As a side effect of this fix,
	get_lists() may return undef.

	[12047] src/bin/create_db.Oracle, src/bin/create_db.Sybase,
	src/lib/Sympa/DatabaseDescription.pm, src/lib/Sympa/Log.pm,
	src/lib/Sympa/Upgrade.pm:  If multiple instances of sympa_msg.pl are
	running, i.e. incoming_max_count parameter is greater than 1, adding
	information to stat_table occasionally fails.  Probably because forked
	instances can share random seed used to generate primary key.
	Fixed by removing id_stat field from stat_table: It was not used
	anyway.

6.2b.8		March 06, 2015

  Features:

	[12026] default/web_tt2/home.tt2:  News about list activities and
	potential user errors are displayed at the home page.



  Changes:

	[12029] default/web_tt2/news.tt2, src/cgi/wwsympa.fcgi.in:  Impoving
	the details given to user when errors are reported to him.

  Bug fixes:

	[12027] src/cgi/wwsympa.fcgi.in, src/lib/Sympa/List.pm:  When
	displaying user errors; errors were displayed even if the user was not
	concerned by the error. Fixed by getting the actual user data.

	[12025] src/lib/Sympa/Upgrade.pm:  When upgrading from pre-beta
	version, archives parameters are renamed. The upgrade process crashed
	on mail_access sub-parameters because they were set as a string instead
	of as a scenario. Simply setting them as a hash containing a 'name'
	entry whose value was the scenarion name was not enough: it led to
	totally messed up archive parameters. Finally fixed by loading the
	mail_access scenario to initialize the mail_access sub-parameter.

6.2b.7		March 04, 2015

  Changes:

	[12004] src/lib/Sympa/Upgrade.pm:  If you upgrade from Sympa 6.2a or
	6.2b, statistics are cleared, because older information is no longer
	usable.

6.2b.6		February 27, 2015

  Changes:

	[11995] src/lib/Sympa/Scenario.pm:  When scenario debug log is
	activated, the scenario file used is also logged.

	[11985] src/lib/Sympa/ListDef.pm, src/libexec/ldap_alias_manager.pl.in:
	 Default cryptographic protocol of LDAP over SSL/TLS became TLSv1. 
	Note that SSLv2 and SSLv3 are no longer recommended due to their
	weakness.
	- List config: The default of "ssl_version" parameters in
	"include_ldap" paragraph and so on became "tlsv1".
	- ldap_alias_manager.pl: New parameter "ldap_ssl_version" to specify
	protocol version which defaults to "tlsv1" was introduced.  On earlier
	releases it was fixed to "sslv3".

  Bug fixes:

	[11994] src/lib/Sympa/Family.pm:  When instantiating a family, Sympa
	would crash if no list is found in the XML file. Preventing it by
	setting the list of lists to create as an empty list. Having an empty
	XML file can be a reasonnable (yet crude) mean to close all lists in a
	family.

	[11989] src/lib/Sympa/Family.pm:  Several bugs made family
	instantiation crash sympa.

	[11988] src/lib/Sympa/List.pm:	When using DB list cache, all lists
	would be returned when queryinf the lists from a family, leading to the
	possibility that ALL lists not belonging to a family would be closed if
	close_unkown option was used.

6.2b.5		February 25, 2015

  Features:

	[11969] src/lib/Sympa/ConfDef.pm, src/sbin/sympa_msg.pl.in: 
	[experimental] Multiple instances of sympa_msg.pl, daemon processing
	incoming message spool, may be run.  Multiplicity is controlled by
	"incoming_max_count" config parameter: By default single instance will
	run.



  Changes:

	[11978] src/bin/mod2html.pl.in, src/bin/testlogs.pl.in,
	src/cgi/sympa_soap_server.fcgi.in, src/cgi/wwsympa.fcgi.in,
	src/lib/Log.pm, src/libexec/alias_manager.pl.in,
	src/sbin/archived.pl.in, src/sbin/bounced.pl.in, src/sbin/bulk.pl.in,
	src/sbin/sympa.pl.in, src/sbin/sympa_automatic.pl.in,
	src/sbin/sympa_msg.pl.in, src/sbin/sympa_newaliases.pl.in,
	src/sbin/task_manager.pl.in:  A few service names in system log and
	database logs were changed.  Now they are taken from the name of each
	program:
	- sympa_automatic.pl : "sympa_automatic" (formerley "sympa")
	- sympa_msg.pl : "sympa_msg" (formerly "sympa")
	Exception:
	- sympa_soap_server.fcgi : "sympasoap" (formerly "soap")
	ToDo: sympa_soap_server.fcgi would be renamed.


src/smtpc/smtpc.c
/* $Id$ */
/*
 * Sympa - SYsteme de Multi-Postage Automatique
 *
 * Copyright (c) 1997, 1998, 1999 Institut Pasteur & Christophe Wolfhugel
 * Copyright (c) 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005,
 * 2006, 2007, 2008, 2009, 2010, 2011 Comite Reseau des Universites
 * Copyright (c) 2011, 2012, 2013, 2014, 2015 GIP RENATER
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/*
 * smtpc was originally written by IKEDA Soji <ikeda@conversion.co.jp>
 * for Sympa project.
 *
 * 2015-05-17 IKEDA Soji: Initial checkin to source repository.
 */

#include "config.h"
#include <signal.h>
#include <stdio.h>
#include <stdarg.h>
#include <stdlib.h>
#include <string.h>
#include <sysexits.h>
#include "sockstr.h"
#include "utf8.h"

#define SMTPC_BUFSIZ (8192)

#define SMTPC_ERR_SOCKET (-1)
#define SMTPC_ERR_PROTOCOL (-2)
#define SMTPC_ERR_SUBMISSION (-3)
#define SMTPC_ERR_UNKNOWN (-4)

#define SMTPC_7BIT (0)
#define SMTPC_8BIT (1)
#define SMTPC_UTF8 (2)

#define SMTPC_PROTO_ESMTP (1)
#define SMTPC_PROTO_LMTP (1 << 1)

#define SMTPC_EXT_8BITMIME (1)
#define SMTPC_EXT_AUTH (1 << 1)
#define SMTPC_EXT_DSN (1 << 2)
#define SMTPC_EXT_PIPELINING (1 << 3)
#define SMTPC_EXT_SIZE (1 << 4)
#define SMTPC_EXT_SMTPUTF8 (1 << 5)
#define SMTPC_EXT_STARTTLS (1 << 6)

#define SMTPC_NOTIFY_NEVER (1)
#define SMTPC_NOTIFY_SUCCESS (1 << 1)
#define SMTPC_NOTIFY_FAILURE (1 << 2)
#define SMTPC_NOTIFY_DELAY (1 << 3)

static char buf[SMTPC_BUFSIZ];
static sockstr_t *sockstr;

static struct {
    int dump;
    int verbose;
    unsigned int protocol;
    char *myname;
    char *nodename;
    char *servname;
    char *path;
    char *sender;
    unsigned int notify;
    char *envid;
    int smtputf8;
} options = {
0, 0, 0, NULL, NULL, NULL, NULL, NULL, 0, NULL, 0};

static struct {
    char **recips;
    int recipnum;
    char *buf;
    size_t buflen;
    size_t size;
    int envfeature;
    int headfeature;
    int bodyfeature;
} message = {
NULL, 0, NULL, 0, 0, 0, 0};

static struct {
    char *buf;
    size_t buflen;
    unsigned long extensions;
} server = {
NULL, 0, 0};

static char *encode_xtext(unsigned char *str)
{
    unsigned char *p;
    char *encbuf, *q;
    size_t enclen = 0;

    p = str;
    while (*p != '\0') {
	if (*p == '+' || *p == '=')
	    enclen += 3;
	else if (33 <= *p && *p <= 126)
	    enclen++;
	else
	    enclen += 3;
	p++;
    }

    encbuf = malloc(enclen + 1);
    if (encbuf == NULL)
	return NULL;

    p = str;
    q = encbuf;
    while (*p != '\0') {
	if (*p == '+' || *p == '=')
	    q += sprintf(q, "+%02X", (unsigned int) *p);
	else if (33 <= *p && *p <= 126)
	    *q++ = *p;
	else
	    q += sprintf(q, "+%02X", (unsigned int) *p);
	p++;
    }
    *q = '\0';

    return encbuf;
}

static int parse_options(int *argcptr, char ***argvptr)
{
    int argc = *argcptr;
    char **argv = *argvptr;

    size_t i;
    char *arg, *p;

    options.dump = 0;
    options.verbose = 0;
    options.protocol = 0;
    options.myname = "localhost";
    options.nodename = NULL;
    options.servname = "25";
    options.sender = NULL;
    options.notify = 0;
    options.envid = NULL;
    options.smtputf8 = 0;

    for (i = 1; i < argc && argv[i] != NULL; i++) {
	arg = argv[i];

	if (arg[0] != '-')
	    break;
	else if (arg[0] == '-' && arg[1] == '-') {
	    if (arg[2] == '\0') {
		i++;
		break;
	    } else if (strcmp(arg, "--dump") == 0)
		options.dump++;
	    else if (strcmp(arg, "--esmtp") == 0 && i + 1 < argc) {
		if (options.protocol != 0) {
		    fprintf(stderr, "Multiple servers are specified\n");
		    return -1;
		}
		options.protocol = SMTPC_PROTO_ESMTP;
		arg = argv[++i];
		if (arg[0] == '[') {
		    p = options.nodename = arg + 1;
		    while (*p != '\0' && *p != ']')
			p++;
		    if (*p == ']' && options.nodename < p)
			*p++ = '\0';
		    else {
			fprintf(stderr, "Malformed host \"%s\"\n", arg);
			return -1;
		    }

		    if (*p == ':' && ++p != '\0')
			options.servname = p;
		    else if (*p != '\0') {
			fprintf(stderr, "Malformed port \"%s\"\n", p);
			return -1;
		    }
		} else {
		    p = options.nodename = arg;
		    while (*p != '\0' && *p != ':')
			p++;
		    if (*p == ':' && options.nodename < p)
			*p++ = '\0';

		    if (*p != '\0')
			options.servname = p;
		}
	    } else if (strcmp(arg, "--iam") == 0 && i + 1 < argc)
		options.myname = argv[++i];
	    else if (strcmp(arg, "--lmtp") == 0 && i + 1 < argc) {
		if (options.protocol != 0) {
		    fprintf(stderr, "Multiple servers are specified\n");
		    return -1;
		}
		options.protocol = SMTPC_PROTO_LMTP;
		options.path = argv[++i];
	    } else if (strcmp(arg, "--smtputf8") == 0)
		options.smtputf8 = 1;
	    else if (strcmp(arg, "--verbose") == 0)
		options.verbose++;
	}

	switch (arg[1]) {
	case 'f':
	    if (options.sender != NULL) {
		fprintf(stderr, "Multiple senders are specified\n");
		return -1;
	    }
	    if (arg[2] == '\0' && i + 1 < argc)
		options.sender = argv[++i];
	    else if (arg[2] != '\0')
		options.sender = arg + 2;
	    else
		goto parse_options_novalue;

	    if (strcmp(options.sender, "<>") == 0)
		options.sender += 2;

	    break;

	case 'N':
	    if (arg[2] == '\0' && i + 1 < argc)
		p = argv[++i];
	    else if (arg[2] != '\0')
		p = arg + 2;
	    else
		goto parse_options_novalue;

	    while (*p != '\0') {
		char word[29], *wp;

		wp = word;
		while (*p == '\t' || *p == ' ' || *p == ',')
		    p++;
		if (*p == '\0')
		    break;

		while (*p != '\0' && *p != '\t' && *p != ' ' && *p != ','
		       && wp - word + 1 < sizeof(word))
		    if ('a' <= *p && *p <= 'z')
			*wp++ = *p++ + ('A' - 'a');
		    else
			*wp++ = *p++;
		*wp = '\0';

		if (strcmp(word, "NEVER") == 0) {
		    options.notify |= SMTPC_NOTIFY_NEVER;
		} else if (strcmp(word, "SUCCESS") == 0)
		    options.notify |= SMTPC_NOTIFY_SUCCESS;
		else if (strcmp(word, "FAILURE") == 0)
		    options.notify |= SMTPC_NOTIFY_FAILURE;
		else if (strcmp(word, "DELAY") == 0)
		    options.notify |= SMTPC_NOTIFY_DELAY;
		else {
		    fprintf(stderr, "Unknown NOTIFY keyword \"%s\"\n",
			    word);
		    return -1;
		}

		if (options.notify & SMTPC_NOTIFY_NEVER &&
		    options.notify & ~SMTPC_NOTIFY_NEVER) {
		    fprintf(stderr,
			    "NEVER keyword must not appear with other keywords\n");
		    return -1;
		}
	    }
	    break;

	case 'V':
	    if (arg[2] == '\0' && i + 1 < argc)
		options.envid = argv[++i];
	    else if (arg[2] != '\0')
		options.envid = arg + 2;
	    else
		goto parse_options_novalue;

	    p = options.envid;
	    while (*p != '\0')
		if (32 <= *p && *p <= 126)
		    p++;
		else {
		    fprintf(stderr,
			    "ENVID contains illegal character \\x%02X\n",
			    *p);
		    return -1;
		}

	    break;

	default:
	    break;

	  parse_options_novalue:
	    fprintf(stderr, "No value for option \"%s\"\n", arg);
	    return -1;
	}
    }

    if (options.protocol & (SMTPC_PROTO_ESMTP | SMTPC_PROTO_LMTP) == 0) {
	fprintf(stderr, "Either --esmtp or --lmtp option must be given\n");
	return -1;
    }
    if (options.protocol & SMTPC_PROTO_ESMTP && options.nodename == NULL
	|| options.protocol & SMTPC_PROTO_LMTP && options.path == NULL) {
	fprintf(stderr, "Nodename nor path is not specified\n");
	return -1;
    }
    if (options.sender == NULL) {
	fprintf(stderr, "Envelope sender is not specified\n");
	return -1;
    }

    *argcptr -= i;
    *argvptr += i;

    return 0;
}

static int check_utf8_address(char *addrbuf)
{
    size_t len;
    ssize_t rs;

    len = strlen(addrbuf);
    if (len == 0)
	return SMTPC_7BIT;

    rs = utf8_check((unsigned char *) addrbuf, len);
    if (rs < 0)
	return SMTPC_7BIT;
    else if (rs < len)
	return SMTPC_8BIT;
    else
	return SMTPC_UTF8;
}

static int read_envelope(char *sender, size_t recipnum, char **recips)
{
    char **pp, **end;

    if (recipnum <= 0) {
	fprintf(stderr, "No recipients are specified\n");
	return -1;
    }

    message.recipnum = recipnum;
    message.recips = recips;

    /*
     * Check feature of sender.
     */
    message.envfeature = check_utf8_address(sender);

    /*
     * Check feature of recipients.
     */
    end = recips + recipnum;
    for (pp = recips;
	 message.envfeature != SMTPC_8BIT && pp < end && *pp != NULL; pp++)
	switch (check_utf8_address(*pp)) {
	case SMTPC_8BIT:
	    message.envfeature = SMTPC_8BIT;
	    break;
	case SMTPC_UTF8:
	    message.envfeature = SMTPC_UTF8;
	    break;
	default:
	    break;
	}

    return 0;
}

static ssize_t read_message(void)
{
    size_t cr;
    ssize_t rs;
    char *newbuf, *p, *end;

    while (1) {
	rs = fread(buf, 1, SMTPC_BUFSIZ, stdin);
	if (rs == 0)
	    break;

	if (message.buf == NULL) {
	    message.buf = malloc(rs + 1);
	    if (message.buf == NULL)
		return -1;
	} else {
	    newbuf = realloc(message.buf, message.buflen + rs + 1);
	    if (newbuf == NULL)
		return -1;
	    message.buf = newbuf;
	}
	memcpy(message.buf + message.buflen, buf, rs);
	message.buflen += rs;
	message.buf[message.buflen] = '\0';

	if (rs < SMTPC_BUFSIZ)
	    break;
    }
    if (feof(stdin)) {
	if (fclose(stdin) != 0)
	    return -1;
    } else {
	fclose(stdin);
	return -1;
    }

    /*
     * Check message features:
     * - Feature of message header.
     * - Feature of message body.
     * - Estimated size of the message considering newlines.
     */

    cr = 0;
    message.headfeature = SMTPC_7BIT;
    message.bodyfeature = SMTPC_7BIT;
    if (0 < message.buflen) {
	end = message.buf + message.buflen;

	for (p = message.buf; p < end; p++)
	    if (*p == '\n') {
		if (p == message.buf || p[-1] != '\r')
		    cr++;

		if (p[1] == '\n' || p[1] == '\r' && p[2] == '\n') {
		    p++;
		    break;
		}
	    }
	rs = utf8_check((unsigned char *) message.buf, p - message.buf);
	if (rs < 0)
	    message.headfeature = SMTPC_7BIT;
	else if (rs < p - message.buf)
	    message.headfeature = SMTPC_8BIT;
	else
	    message.headfeature = SMTPC_UTF8;

	for (; p < end; p++)
	    if (*p & 0x80) {
		message.bodyfeature = SMTPC_8BIT;
		p++;
		break;
	    } else if (*p == '\n' && p[-1] != '\r')
		cr++;
	for (; p < end; p++)
	    if (*p == '\n' && p[-1] != '\r')
		cr++;

	if (end[-1] == '\r')
	    cr++;
	else if (end[-1] != '\n')
	    cr += 2;
    } else
	cr = 2;
    message.size = message.buflen + cr;

    return message.buflen;
}

static int dialog(int timeout, const char *format, ...)
{
    va_list ap;
    ssize_t rs;

    sockstr->timeout = timeout;

    if (format != NULL && *format != '\0') {
	if (options.dump) {
	    fprintf(stderr, "C: ");
	    va_start(ap, format);
	    vfprintf(stderr, format, ap);
	    va_end(ap);
	}
	va_start(ap, format);
	rs = sockstr_vprintf(sockstr, format, ap);
	va_end(ap);
	if (rs < 0)
	    return -1;
    }

    rs = sockstr_getstatus(sockstr, &server.buf, &server.buflen);
    if (rs <= 0)
	return -1;

    if (options.dump)
	fprintf(stderr, "%s", server.buf);
    return server.buf[0];
}

static int datasend(int timeout)
{
    ssize_t rs;

    sockstr->timeout = timeout;

    if (options.dump)
	fprintf(stderr, "C: (MESSAGE)\r\nC: .\r\n");
    if (sockstr_putdata
	(sockstr, message.buf, message.buflen, ".\r\n", 1, 1) < 0)
	return -1;

    rs = sockstr_getstatus(sockstr, &server.buf, &server.buflen);
    if (rs <= 0)
	return -1;

    if (options.dump)
	fprintf(stderr, "%s", server.buf);
    return server.buf[0];
}

static void parse_extensions(void)
{
    char *p = server.buf;
    unsigned long extensions = 0L;
    char word[512], *wp;

    while (*p != '\n' && *p != '\0')
	p++;
    if (*p == '\n')
	p++;
    while (*p != '\0') {
	if (p[0] && p[1] && p[2] && p[3]) {
	    p += 4;
	    while (*p == '\t' || *p == ' ' || *p == '-')
		p++;
	    if (*p == '\0')
		break;

	    wp = word;
	    while (wp - word + 1 < sizeof(word) &&
		   (*p == '-' || '0' <= *p && *p <= '9' ||
		    'A' <= *p && *p <= 'Z' || 'a' <= *p && *p <= 'z'))
		if ('a' <= *p && *p <= 'z')
		    *wp++ = *p++ + ('A' - 'a');
		else
		    *wp++ = *p++;
	    *wp = '\0';
	    if (strcmp(word, "8BITMIME") == 0)
		extensions |= SMTPC_EXT_8BITMIME;
	    else if (strcmp(word, "AUTH") == 0)
		extensions |= SMTPC_EXT_AUTH;
	    else if (strcmp(word, "DSN") == 0)
		extensions |= SMTPC_EXT_DSN;
	    else if (strcmp(word, "PIPELINING") == 0)
		extensions |= SMTPC_EXT_PIPELINING;
	    else if (strcmp(word, "SIZE") == 0)
		extensions |= SMTPC_EXT_SIZE;
	    else if (strcmp(word, "SMTPUTF8") == 0)
		extensions |= SMTPC_EXT_SMTPUTF8;
	    else if (strcmp(word, "STARTTLS") == 0)
		extensions |= SMTPC_EXT_STARTTLS;
	}

	while (*p != '\n' && *p != '\0')
	    p++;
	if (*p == '\n')
	    p++;
    }

    server.extensions = extensions;
}

static ssize_t transaction(void)
{
    ssize_t sent = 0;
    char *hello;
    char *ext_8bitmime, ext_envid[108], ext_notify[37], ext_size[27],
	*ext_smtputf8;
    int i;

    ext_8bitmime = "";
    *ext_envid = '\0';
    *ext_notify = '\0';
    *ext_size = '\0';
    ext_smtputf8 = "";

    if (options.protocol & SMTPC_PROTO_ESMTP)
	hello = "EHLO";
    else if (options.protocol & SMTPC_PROTO_LMTP)
	hello = "LHLO";
    else
	return SMTPC_ERR_UNKNOWN;
    switch (dialog(300, "%s %s\r\n", hello, options.myname)) {
    case '2':
	break;
    case '4':
    case '5':
	return 0;
    case -1:
	return SMTPC_ERR_SOCKET;
    default:
	return SMTPC_ERR_PROTOCOL;
    }
    parse_extensions();

    if (server.extensions & SMTPC_EXT_8BITMIME &&
	(message.headfeature != SMTPC_7BIT
	 || message.bodyfeature != SMTPC_7BIT))
	ext_8bitmime = " BODY=8BITMIME";

    if (server.extensions & SMTPC_EXT_DSN) {
	if (options.envid != NULL && options.envid[0] != '\0') {
	    unsigned char *encbuf;

	    encbuf = encode_xtext((unsigned char *) options.envid);
	    if (encbuf == NULL) {
		perror("transaction");
		return SMTPC_ERR_UNKNOWN;
	    }
	    snprintf(ext_envid, sizeof(ext_envid), " ENVID=%s", encbuf);
	    free(encbuf);
	}

	if (options.notify) {
	    unsigned int mask;

	    for (mask = 1; mask < (1 << 4); mask <<= 1) {
		if (options.notify & mask) {
		    if (*ext_notify == '\0')
			strcat(ext_notify, " NOTIFY=");
		    else
			strcat(ext_notify, ",");

		    switch (mask) {
		    case SMTPC_NOTIFY_NEVER:
			strcat(ext_notify, "NEVER");
			break;
		    case SMTPC_NOTIFY_SUCCESS:
			strcat(ext_notify, "SUCCESS");
			break;
		    case SMTPC_NOTIFY_FAILURE:
			strcat(ext_notify, "FAILURE");
			break;
		    case SMTPC_NOTIFY_DELAY:
			strcat(ext_notify, "DELAY");
			break;
		    }
		}
	    }			/* for (mask ...) */
	}			/* if (options.notify & mask) */
    }

    if (server.extensions & SMTPC_EXT_SIZE)
	snprintf(ext_size, sizeof(ext_size), " SIZE=%lu",
		 (unsigned long) message.size);

    if (server.extensions & SMTPC_EXT_SMTPUTF8 &&
	options.smtputf8 &&
	(message.envfeature == SMTPC_UTF8
	 && message.headfeature != SMTPC_8BIT
	 || message.envfeature != SMTPC_8BIT
	 && message.headfeature == SMTPC_UTF8))
	ext_smtputf8 = " SMTPUTF8";

    switch (dialog(300, "MAIL FROM:<%s>%s%s%s%s\r\n", options.sender,
		   ext_8bitmime, ext_envid, ext_size, ext_smtputf8)) {
    case '2':
	break;
    case '4':
    case '5':
	return 0;
    case -1:
	return SMTPC_ERR_SOCKET;
    default:
	return SMTPC_ERR_PROTOCOL;
    }

    for (i = 0; i < message.recipnum; i++)
	switch (dialog
		(300, "RCPT TO:<%s>%s\r\n", message.recips[i],
		 ext_notify)) {
	case '2':
	    sent++;
	    break;
	case '4':
	case '5':
	    return 0;
	case -1:
	    return SMTPC_ERR_SOCKET;
	default:
	    return SMTPC_ERR_PROTOCOL;
	}

    switch (dialog(120, "DATA\r\n")) {
    case '3':
	break;
    case '4':
    case '5':
	return 0;
    case -1:
	return SMTPC_ERR_SOCKET;
    default:
	return SMTPC_ERR_PROTOCOL;
    }

    switch (datasend(600)) {
    case '2':
	return sent;
    case '5':
	return 0;
    case -1:
	return SMTPC_ERR_SOCKET;
    default:
	return SMTPC_ERR_PROTOCOL;
    }
}

static ssize_t session()
{
    ssize_t rs;

    switch (dialog(300, NULL)) {
    case '2':
	break;
    case '4':
    case '5':
	return SMTPC_ERR_SUBMISSION;
    case -1:
	return SMTPC_ERR_SOCKET;
    default:
	return SMTPC_ERR_PROTOCOL;
    }

    rs = transaction();
    if (rs == 0)
	return SMTPC_ERR_SUBMISSION;
    return rs;
}

int main(int argc, char *argv[])
{
    ssize_t rs;

    if (parse_options(&argc, &argv) < 0)
	exit(EX_USAGE);
    if (read_envelope(options.sender, argc, argv) < 0)
	exit(EX_USAGE);
    if (read_message() < 0) {
	perror("read_message");
	if (message.buf != NULL)
	    free(message.buf);
	exit(EX_OSERR);
    }

    if (options.protocol & SMTPC_PROTO_ESMTP)
	sockstr = sockstr_new(options.nodename, options.servname, NULL);
    else if (options.protocol & SMTPC_PROTO_LMTP)
	sockstr = sockstr_new(NULL, NULL, options.path);
    else
	sockstr = NULL;
    if (sockstr == NULL) {
	perror("sockstr_new");
	if (message.buf != NULL)
	    free(message.buf);
	exit(EX_OSERR);
    }
    sockstr->timeout = 300;

    signal(SIGPIPE, SIG_IGN);

    if (sockstr_connect(sockstr) < 0) {
	fprintf(stderr, "error: %s\n", sockstr_errstr(sockstr));
	sockstr_destroy(sockstr);
	if (message.buf != NULL)
	    free(message.buf);
	exit(EX_IOERR);
    }

    rs = session();
    if (message.buf != NULL)
	free(message.buf);

    if (options.verbose && server.buf != NULL && 0 < server.buflen)
	fputs(server.buf, stdout);

    if (rs < 0) {
	switch (rs) {
	case SMTPC_ERR_SOCKET:
	    fprintf(stderr, "Socket error: %s\n", sockstr_errstr(sockstr));
	    break;
	case SMTPC_ERR_PROTOCOL:
	    fprintf(stderr, "Unexpected response: %s", server.buf);
	    break;
	case SMTPC_ERR_SUBMISSION:
	    dialog(10, "QUIT\r\n");
	    break;
	default:
	    fprintf(stderr, "Unknown error %ld\n", (long) rs);
	    break;
	}

	sockstr_destroy(sockstr);
	if (server.buf != NULL)
	    free(server.buf);
	exit(rs);
    }

    /* entirely or partially sent */
    dialog(10, "QUIT\r\n");
    sockstr_destroy(sockstr);
    free(server.buf);
    exit(EX_OK);
}
